Added the implementation of the Edmond Karp algorithm along with test cases

mapcrafter2048 opened the pull request:
Added the implementation of the Edmond Karp algorithm along with test cases

appgurueu commented on graph/edmondkarp.ts:
```diff
@@ -0,0 +1,76 @@
+/**
+ * @function edmondkarp
```
```suggestion

```

appgurueu commented on graph/edmondkarp.ts:
```diff
@@ -0,0 +1,76 @@
+/**
+ * @function edmondkarp
+ * @description Compute the maximum flow from a source node to a sink node. The input graph is in adjacency list form. It is a multidimensional array of edges. graph[i] holds the edges for the i'th node. Each edge is a 3-tuple where the 0'th item is the destination node, the 1'th item is the edge weight, and the 2'nd item is the edge capacity.
```
By "weight" you mean flow? How is this used?

mapcrafter2048 replied:
yeah I understood that is my conceptual error of including weight

appgurueu commented on graph/edmondkarp.ts:
```diff
@@ -0,0 +1,76 @@
+ * @description Compute the maximum flow from a source node to a sink node. The input graph is in adjacency list form. It is a multidimensional array of edges. graph[i] holds the edges for the i'th node. Each edge is a 3-tuple where the 0'th item is the destination node, the 1'th item is the edge weight, and the 2'nd item is the edge capacity.
+ * @Complexity_Analysis
+ * Time complexity: O(V*E^2) where V is the number of vertices and E is the number of edges
+ * Space Complexity: O(V) where V is the number of vertices
```
This is not true: You initialize O(V²) space matrices below.

appgurueu commented on graph/edmondkarp.ts:
```diff
@@ -0,0 +1,76 @@
+
+    // Build capacity matrix 
+    for (let u = 0; u < n; u++) {
+        for (const [v, , cap] of graph[u]) {
```
It looks like the weight is entirely unused; why is it present at all then?

mapcrafter2048 replied:
sorry for my mistake this is the thing I overlooked. I am truly sorry for this mistake

appgurueu commented on graph/edmondkarp.ts:
```diff
@@ -0,0 +1,76 @@
+        // Find an augmenting path from source to sink by doing a BFS traversal
+        while (queue.length > 0) {
+            // Dequeue
+            const u = queue.shift()!;
```
As said since this is linear time you can't maintain a good time complexity this way. Either use a Queue implementation from this repo or implement the BFS as levelorder by using two level arrays, one for the current and one for the next level, and swapping them.

mapcrafter2048 replied:
I have made an attempt to fix this issue in my new pull request

appgurueu commented on graph/edmondkarp.ts:
```diff
@@ -0,0 +1,76 @@
+            // Dequeue
+            const u = queue.shift()!;
+            // Enqueue all adjacent unvisited vertices with available capacity
+            for (let v = 0; v < n; v++) {
```
You will want to store the residual graph in adjacency list representation as well. Then you should be iterating over the outgoing edges in the residual graph here.

This is needlessly inefficient. It makes your BFS $O(V^2)$ rather than $O(E)$.

appgurueu commented on graph/edmondkarp.ts:
```diff
@@ -0,0 +1,76 @@
+                    queue.push(v);
+                    visited[v] = true;
+                    parent[v] = u;
+                    // If we reach the sink, we have found the augmenting path
```
```suggestion
                    // If we reach the sink, we have found an augmenting path
```

appgurueu commented on graph/edmondkarp.ts:
```diff
@@ -0,0 +1,76 @@
+    };
+
+    let maxFlow = 0;
+    const parent = Array(n).fill(-1);
```
This is TypeScript: I think it would be more idiomatic to use `null` here to signal absence of a value.

appgurueu commented on maths/bisection_method.ts:
```diff
@@ -0,0 +1,27 @@
+/**
```
Please get rid of these unrelated changes in this PR.

appgurueu replied:
I'm afraid this isn't resolved. The changes are still there.

appgurueu replied:
Still there. Maybe you forgot to push a commit that gets rid of these changes?

mapcrafter2048 commented:
I am extremely sorry for all the mistakes I have done and submitted the code for a review with twice checking it my self

mapcrafter2048 commented:
I have tried implementing the changes suggest by @appgurueu 
-removed the weight as it was not used here
-corrected the documentation for the space complexity form O(V) -> O9=(V^2)
-Implemented BFS using two level arrays (currentLevel and nextLevel) to maintain level-order traversal, avoiding the inefficient queue.shift() operation.
-Used an adjacency list representation for the residual graph to make the iteration over outgoing edges
-Changed the parent array initialization to null

appgurueu commented on graph/edmondkarp.ts:
```diff
@@ -0,0 +1,96 @@
+ * @see https://en.wikipedia.org/wiki/Edmonds%E2%80%93Karp_algorithm
+ */
+
+function edmondkarp(graph: [number, number][][], source: number, sink: number): number {
```
This should be called `edmondsKarp`, and it needs an `export` (`default`). The file should also be called `edmond_karp.ts`.

mapcrafter2048 reviewed (commented):
implement the changes suggest by you

appgurueu commented on graph/edmonds_karp.ts:
```diff
@@ -0,0 +1,96 @@
+    parent[source] = null
+
+    while (queue.length > 0) {
+      const u = queue.shift()!
```
You're still $O(n)$ shifting here. This is still needlessly inefficient. As said, please use one of our efficient queue implementations such as https://github.com/TheAlgorithms/TypeScript/blob/master/data_structures/queue/stack_queue.ts. Or do it as a levelorder where you build the next level from the previous one, then swap.

appgurueu reviewed (approved):
Should be asymptotically fine now. Looks correct.
